components/hab/src/command/pkg/binlink.rs

```diff
@@ -135,89 +141,159 @@ pub fn binlink_all_in_pkg(ui: &mut UI,
-        path.set_extension("bat");
-
-        Ok(Binlink { dest: path,
-                     src:  src.as_ref().to_path_buf(), })
+    fn binstub_path(src: &Path, dest_dir: &Path) -> Result<PathBuf> {
+        #[cfg(windows)]
+        {
+            let bin_name = match src.file_stem() {
+                Some(name) => name,
+                None => return Err(Error::CannotParseBinlinkSource(src.to_path_buf())),
+            };
+            let mut path = dest_dir.join(bin_name);
+            path.set_extension("bat");
+            Ok(path)
+        }
+
+        #[cfg(unix)]
+        match src.file_name() {
+            Some(name) => Ok(dest_dir.join(name)),
+            None => Err(Error::CannotParseBinlinkSource(src.to_path_buf())),
+        }
     }
 
-    pub fn from_file<T: AsRef<Path>>(path: T) -> Result<Self> {
-        use toml::Value::Table;
+    fn stub_template(&self, env: HashMap<String, String>) -> Result<String> {
+        let mut exports = String::new();
+        for (key, mut value) in env.into_iter() {
+            if key == "PATH" {
+                let mut paths: Vec<PathBuf> = env::split_paths(&value).collect();
+                paths.push(PathBuf::from(Self::env_var("PATH")));
+                value =
+                    env::join_paths(paths)?.into_string()
+                                           .expect("Failed to convert binlink PATH to utf8 string");
+            }
+            exports.push_str(&format!("{} {}={}\n", SET_OR_EXPORT, key, value));
+        }
 
-        let file = File::open(&path)?;
-        for line in BufReader::new(file).lines() {
-            let ln = line?;
-            if ln.to_uppercase().starts_with(BAT_COMMENT_MARKER) {
-                let (_, rest) = ln.split_at(BAT_COMMENT_MARKER.len());
-                if let Ok(Table(toml_exp)) = rest.parse() {
-                    if let Some(src) = toml_exp.get("source") {
-                        if let Some(val) = src.as_str() {
-                            return Ok(Binlink { dest: path.as_ref().to_path_buf(),
-                                                src:  PathBuf::from(val), });
-                        }
-                    }
+        #[cfg(windows)]
+        {
+            Ok(format!(include_str!("../../../static/template_binstub.\
+                                     bat"),
+                       src = self.src.display(),
+                       env = exports))
+        }
+
+        #[cfg(unix)]
+        {
+            // We need to prevent the binstub from using itself as an interpreter.
+            // On linux we use `/bin/sh` as the binstub interpreter and if this package
+            // includes a `sh` binary, we risk a circularly linked interpretter. We can
+            // detect if the link path is the same as the interpreter and if so, use the
+            // src binary as the interpreter.
+            let interpreter = {
+                if self.default_interpreter == self.dest {
```

Comment:
I wonder if we need to [`canonicalize`](http://doc.rust-lang.org/1.36.0/std/path/struct.Path.html#method.canonicalize) here. It's not uncommon for `/bin/sh` to be a link to some other interpreter. On my Ubuntu system: for instance:
```
➤ ls -l /bin/sh
lrwxrwxrwx 1 root root 4 Dec 13  2017 /bin/sh -> dash*
```